src/java_bytecode/java_bytecode_convert_class.cpp

```diff
@@ -85,6 +85,70 @@ class java_bytecode_convert_classt:public messaget
   static void add_array_types(symbol_tablet &symbol_table);
 };
 
+/// Auxiliary function to extract the generic superclass reference from the
+/// class signature. If the signature is empty or the superclass is not generic
+/// it returns empty.
+/// Example:
+/// - class: A<T> extends B<T, Integer> implements C, D<T>
+/// - signature: <T:Ljava/lang/Object;>B<TT;Ljava/lang/Integer;>;LC;LD<TT;>;
+/// - superclass reference: B<TT;Ljava/lang/Integer;>;
+/// \param signature Signature of the class
+/// \return Reference of the generic superclass, or empty if the superclass
+/// is not generic
+static optionalt<std::string>
+extract_generic_superclass_reference(const optionalt<std::string> &signature)
+{
+  if(signature.has_value())
+  {
+    // skip the (potential) list of generic parameters at the beginning of the
+    // signature
+    const size_t start =
+      signature.value().front() == '<'
+        ? find_closing_delimiter(signature.value(), 0, '<', '>') + 1
+        : 0;
+
+    // extract the superclass reference
+    const size_t end =
+      find_closing_semi_colon_for_reference_type(signature.value(), start);
+    const std::string superclass_ref =
+      signature.value().substr(start, end - start + 1);
+
+    // if the superclass is generic then the reference is of form
+    // Lsuperclass-name<generic-types;>;
+    if(superclass_ref.substr(superclass_ref.length() - 2) == ">;")
+      return superclass_ref;
+  }
+  return {};
+}
+
+/// Auxiliary function to extract the generic interface reference of an
+/// interface with the specified name from the class signature. If the
+/// signature is empty or the interface is not generic it returns empty.
+/// Example:
+/// - class: A<T> extends B<T, Integer> implements C, D<T>
+/// - signature: <T:Ljava/lang/Object;>B<TT;Ljava/lang/Integer;>;LC;LD<TT;>;
+/// - interface reference for C: LC;
+/// - interface reference for D: LD<TT;>;
+/// \param signature Signature of the class
+/// \param interface_name The interface name
+/// \return Reference of the generic interface, or empty if the interface
+/// is not generic
+static optionalt<std::string> extract_generic_interface_reference(
+  const optionalt<std::string> &signature,
+  const std::string &interface_name)
+{
+  if(signature.has_value())
+  {
+    const size_t start = signature.value().find("L" + interface_name + "<");
```

Comment:
Could the interface also occur earlier? For example, for `class A extends ArrayList<Closable> implements Closable`, would this find the reference to Closable as an argument to ArrayList rather than the interface as desired?

Reply:
Thank you for pointing this out! Yes, the interface can occur earlier as in the example you gave. However, we are only interested in the reference to the interface if it is generic which is ensured by looking for substring with `<` after the name of the interface. In your example there would be no problem. The problem arises in case like this `class A extends B<Interface<String>> implements Interface<Integer>` where the signature is:
`LB<LInterface<Ljava/lang/String;>;>;LInterface<Ljava/lang/Integer;>;`
The function should return reference `LInterface<Ljava/lang/Integer;>;` but this implementation returns `LInterface<Ljava/lang/String;>;`. I've corrected the code and added unit tests for this case.

Reply:
And this also added unit tests for nested generics in bases which we didn't have before so double thanks :)